models/pull.go

```diff
@@ -270,63 +270,98 @@ func (pr *PullRequest) Merge(doer *User, baseGitRepo *git.Repository) (err error
 		return fmt.Errorf("OpenRepository: %v", err)
 	}
 
-	// Clone base repo.
-	tmpBasePath := path.Join(setting.AppDataPath, "tmp/repos", com.ToStr(time.Now().Nanosecond())+".git")
+	baseRepoPath := pr.BaseRepo.RepoPath()
 
-	if err := os.MkdirAll(path.Dir(tmpBasePath), os.ModePerm); err != nil {
-		return fmt.Errorf("Failed to create dir %s: %v", tmpBasePath, err)
-	}
+	pullHead := fmt.Sprintf("refs/pull/%d/head", pr.Index)
 
-	defer os.RemoveAll(path.Dir(tmpBasePath))
+	// A temporary Git index file we are going to build. The merge commit will be based on it.
+	indexTmpPath := filepath.Join(os.TempDir(), "gitea-"+pr.BaseRepo.Name+"-"+strconv.Itoa(time.Now().Nanosecond()))
+	defer os.Remove(indexTmpPath)
 
-	var stderr string
-	if _, stderr, err = process.GetManager().ExecTimeout(5*time.Minute,
-		fmt.Sprintf("PullRequest.Merge (git clone): %s", tmpBasePath),
-		"git", "clone", baseGitRepo.Path, tmpBasePath); err != nil {
-		return fmt.Errorf("git clone: %s", stderr)
+	// A temporary Git working tree for git-merge-index and git-merge-one-file.
+	workTreeTmpPath, err := ioutil.TempDir("", "gitea-merge-")
```

Comment:
Is not there missing `+pr.BaseRepo.Name+"-"+strconv.Itoa(time.Now().Nanosecond())`?

Reply:
ioutil.TempDir will generate a suffix automatically. See https://golang.org/src/io/ioutil/tempfile.go?s=2138:2195#L66

Reply:
Ah, do you want a different prefix like `pr.BaseRepo.Name+`?

Reply:
>Is not there missing +pr.BaseRepo.Name+"-"+strconv.Itoa(time.Now().Nanosecond())?
This is actually correct despite a bit of inconsistent. I will clean this up a little bit.

The path `indexTmpPath` and `workTreeTmpPath` conrespond to `GIT_INDEX_FILE` and `GIT_WORK_TREE` respectively. I looked at this in a hurry last friday and got confused as well. 😆